tests/suites/hup/tests/storagemigration.js

```diff
@@ -0,0 +1,490 @@
+					true,
+					'All containers should be running on the overlay2 storage-driver',
+				);
+
+				await archiveMigrationLogs(this, test);
+			},
+		},
+		{
+			title: 'Migration should succeed when rollback triggers',
+			run: async function(test) {
+				await this.context.get().hup.initDUT(this, test, this.context.get().link);
+
+				if (!(await shouldMigrate(this, test))) {
+					return; // SKIP
+				}
+
+				const versionBeforeHup = await this.context
+					.get()
+					.worker.getOSVersion(this.context.get().link);
+
+				test.comment(`OS version before HUP: ${versionBeforeHup}`);
+
+				const containersBeforeHup = (await this.context
+					.get()
+					.worker.executeCommandInHostOS(
+						`balena container ls --format '{{.ID}}'`,
+						this.context.get().link,
+					))
+					.trim()
+					.split('\n');
+
+				await this.context
+					.get()
+					.hup.doHUP(
+						this,
+						test,
+						'image',
+						this.context.get().hup.payload,
+						this.context.get().link,
+					);
+
+				// TODO: this copies a large chunk of the rollback-health/openvpn
+				// test case. how useful is this? we just want to check that
+
+				// break openvpn
+				test.comment(`Breaking openvpn to trigger rollback-health...`);
+				await this.context
+					.get()
+					.worker.executeCommandInHostOS(
+						`ln -sf /dev/null $(find /mnt/sysroot/inactive/ | grep "bin/openvpn$")`,
+						this.context.get().link,
+					);
+
+				test.comment(
+					`Pretend VPN was previously active for unmanaged OS suite...`,
+				);
+				await this.context
+					.get()
+					.worker.executeCommandInHostOS(
+						`sed 's/BALENAOS_ROLLBACK_VPNONLINE=0/BALENAOS_ROLLBACK_VPNONLINE=1/' -i /mnt/state/rollback-health-variables && sync -f /mnt/state`,
+						this.context.get().link,
+					);
+
+				await rebootAndWaitForRollbackHealth(this, test);
+
+				test.is(
+					await this.context.get().worker.getOSVersion(this.context.get().link),
+					versionBeforeHup,
+					`The OS version should have reverted to ${versionBeforeHup}`,
+				);
+
+				// 0 means file exists, 1 means file does not exist
+				test.is(
+					await this.context
+						.get()
+						.worker.executeCommandInHostOS(
+							`test -f /mnt/state/rollback-health-triggered ; echo $?`,
+							this.context.get().link,
+						),
+					'0',
+					'There should be a rollback-health-triggered file in the state partition',
+				);
+
+				// 0 means directory exists, 1 means directory does not exist
+				test.is(
+					await this.context.get().worker.executeCommandInHostOS(
+						`test -d /mnt/data/docker/overlay2; echo $?`,
+						this.context.get().link,
+					),
+					'0',
+					'There should be a overlay2 directory under /var/lib/docker',
+				);
+
+				// 0 means we migrated fully, anything else means we didn't
+				const didMigrate = (
+					await this.context.get().worker.executeCommandInHostOS(
+						`journalctl --no-pager -ru balena -t balenad | grep 'Storage migration finished'; echo $?`,
```

Comment:
Did the storage migration run on the current boot after the rollback? Otherwise these logs wouldn't be available on the current boot.

Note that we can't use `-b all` here either because after a rollback the previous OS version may not support the `all` option like we found with RPi2